src/test/java/org/opensearch/knn/index/AdvancedFilteringUseCasesIT.java

```diff
@@ -0,0 +1,449 @@
+
+    private static final String FIELD_NAME_NESTED = "test_nested";
+
+    private static final String FIELD_NAME_VECTOR = "test_vector";
+
+    private static final String PROPERTIES_FIELD = "properties";
+
+    private static final String FILTER_FIELD = "filter";
+
+    private static final String TERM_FIELD = "term";
+
+    private static final int k = 10;
+
+    private static final String FIELD_NAME_METADATA = "parking";
+
+    private static final int NUM_DOCS = 5;
+
+    private static final int DOCUMENT_IN_RESPONSE = 2;
+
+    private static final Float[] QUERY_VECTOR = { 3f };
+
+    private static final List<String> enginesToTest = KNNEngine.getEnginesThatSupportsFilters()
+        .stream()
+        .map(KNNEngine::getName)
+        .collect(Collectors.toList());
+
+    /**
+     * {
+     *     "query": {
+     *         "nested": {
+     *             "path": "test_nested",
+     *             "query": {
+     *                 "knn": {
+     *                     "test_nested.test_vector": {
+     *                         "vector": [
+     *                             3
+     *                         ],
+     *                         "k": 10,
+     *                         "filter": {
+     *                             "nested": {
+     *                                 "path": "test_nested",
+     *                                 "query": {
+     *                                     "term": {
+     *                                         "test_nested.parking": "false"
+     *                                     }
+     *                                 }
+     *                             }
+     *                         }
+     *                     }
+     *                 }
+     *             }
+     *         }
+     *     }
+     * }
+     */
+    @SneakyThrows
+    public void testFiltering_whenNestedKNNAndFilterFieldWithNestedQueries_thenSuccess() {
+        for (final String engine : enginesToTest) {
+            // Set up the index with nested k-nn and metadata fields
+            createKnnIndex(INDEX_NAME, createNestedMappings(1, engine));
+            for (int i = 1; i <= NUM_DOCS; i++) {
+                // making sure that only 2 documents have valid filters
+                final String metadataFieldValue = i % 2 == 0 ? "false" : "true";
+                String doc = NestedKnnDocBuilder.create(FIELD_NAME_NESTED)
+                    .addVectors(FIELD_NAME_VECTOR, new Float[] { (float) i + 1 })
+                    .addVectorWithMetadata(FIELD_NAME_VECTOR, new Float[] { (float) i }, FIELD_NAME_METADATA, metadataFieldValue)
+                    .build();
+                addKnnDoc(INDEX_NAME, String.valueOf(i), doc);
+            }
+            forceMergeKnnIndex(INDEX_NAME);
+
+            // Build the query with both k-nn and filters as nested fields. The filter should also have a nested context
+            final XContentBuilder builder = XContentFactory.jsonBuilder().startObject().startObject(QUERY);
```

Comment:
Can we create a method that return Xcontentbuilder object by passing values. As these work has been repeated other tests also.

Reply:
its not repeated. The queries that are getting created are very different from one another. Abstracting in a function is not possible here.